src/main.cpp

```diff
@@ -6147,9 +6168,28 @@ bool static ProcessMessage(CNode* pfrom, string strCommand, CDataStream& vRecv,
         CXThinBlockTx thinBlockTx;
         vRecv >> thinBlockTx;
 
+        // Message consistency checking
         CInv inv(MSG_XTHINBLOCK, thinBlockTx.blockhash);
+        if (thinBlockTx.vMissingTx.empty() || thinBlockTx.blockhash.IsNull() || pfrom->xThinBlockHashes.size() != pfrom->thinBlock.vtx.size())
+        {
+            {
+                LOCK(cs_vNodes);
```

Comment:
we need this because vnodes is not locked and since we are doing parallel validations we could have contention here .